citesphere/src/main/java/edu/asu/diging/citesphere/web/user/MoveItemsController.java

```diff
@@ -0,0 +1,86 @@
+
+import edu.asu.diging.citesphere.core.exceptions.CannotFindCitationException;
+import edu.asu.diging.citesphere.core.service.ICitationManager;
+import edu.asu.diging.citesphere.core.util.model.ICitationHelper;
+import edu.asu.diging.citesphere.model.bib.ICitation;
+import edu.asu.diging.citesphere.user.IUser;
+import edu.asu.diging.citesphere.web.user.dto.MoveItemsRequest;
+
+@Controller
+public class MoveItemsController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private ICitationManager citationManager;
+    
+    @Autowired
+    private AsyncTaskProcessor<ZoteroUpdateItemsStatuses> asyncTaskProcessor;
+
+    @Autowired
+    private ICitationHelper citationHelper;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/move", method = RequestMethod.POST)
+    public @ResponseBody String moveItemsToCollection(Authentication authentication,
+            @PathVariable("zoteroGroupId") String zoteroGroupId, @RequestBody String itemsData)
+            throws Exception {
+        Gson gson = new Gson();
+        MoveItemsRequest itemsDataDto = gson.fromJson(itemsData, MoveItemsRequest.class);
+        List<ICitation> citations = new ArrayList<>();
+        ICitation citation;
+        for (String key : itemsDataDto.getItemIds()) {
+            try {
+                citation = citationManager.getCitation((IUser) authentication.getPrincipal(), zoteroGroupId, key);
+                citations.add(citation);
+            } catch (CannotFindCitationException e) {
+                logger.error("Cannot find citation.", e);
+                continue;
+            }
+            citationHelper.addCollection(citation, itemsDataDto.getCollectionId(),
+                    (IUser) authentication.getPrincipal());
+        }
+        
+        AsyncTaskResponse<ZoteroUpdateItemsStatuses> asyncResponse;
+        try {
+            asyncResponse = asyncTaskProcessor.submitTask(() -> citationManager.updateCitations((IUser) authentication.getPrincipal(),
+                    zoteroGroupId, citations));
+        } catch (Exception e) {
+            logger.error("Unable to move citations.", e);
+            throw e;
+        }
+                
+        return gson.toJson(asyncResponse, AsyncTaskResponse.class);
+    }
+    
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/move/status", method = RequestMethod.POST)
+    public @ResponseBody AsyncTaskResponse<ZoteroUpdateItemsStatuses> getMoveItemsStatus(Authentication authentication,
+            @PathVariable("zoteroGroupId") String zoteroGroupId, @RequestBody String taskID) throws Exception {
+        return asyncTaskProcessor.getResponse(taskID);
+    }
+    
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/move/cleartask", method = RequestMethod.POST)
+    public @ResponseBody Boolean clearTask(Authentication authentication,
+            @PathVariable("zoteroGroupId") String zoteroGroupId, @RequestBody String taskID) {
```

Comment:
I would put the taskId into the path or as request parameter.

Reply:
and then make the url `.../move/task/{taskId}/clear` (f you make it a path variable). "cleartask" is unnecessarily long.